### PR TITLE
perf(sm80): avoid long-context MQA register spills

### PR DESCRIPTION
**Pull request by fidonan**
## Purpose

On exact SM80 devices, the portable FP8 MQA prefill kernel switched from
`BLOCK_M=16` to `BLOCK_M=64` after the compressed KV sequence exceeded 16K
tokens. For the long-context C4-indexer shapes tested on A100, the M64 kernel
spilled heavily to local memory and caused a sharp prefill slowdown.

This PR keeps M16 for exact SM80 while preserving the existing selector policy
for SM86, SM89, and SM120.

## Scope

This PR contains only the SM80 prefill-kernel change and its reproducibility
material:

- exact-SM80 selector gate;
- selector and numerical-equivalence regressions;
- a same-shape M16/M64 CUDA-event microbenchmark;
- bilingual source and hybrid-deployment guidance;
- measured A100 kernel and end-to-end observations.

It does not add a Router, protocol adapter, service-manager configuration, or
machine-specific deployment data.

The selector change is sufficient to reproduce the M16 kernel behavior. A full
long-running hybrid deployment also requires the runtime/offload stack and the
correctness backports listed in the README; those are deliberately not folded
into this focused performance PR.

This is not a duplicate of #51, which fixes 32-bit paged-MQA block addressing,
or #61, which addresses request-dependent Triton specialization-cache growth.
Neither changes the prefill tile selector.

## Strict kernel benchmark

The benchmark uses one untimed compilation/warm-up phase followed by CUDA-event
timing and reports the median. Each row compares identical inputs and the full
FP32 logits tensors were bitwise equal.

| Raw context | Compressed KV | M16 median | M64 median | Speedup | Spill M16 / M64 |
|---:|---:|---:|---:|---:|---:|
| 98,304 | 24,576 | 40.491 ms | 643.797 ms | 15.90x | 0 / 5,744 B per thread |
| 131,072 | 32,768 | 34.065 ms | 1,003.002 ms | 29.44x | 0 / 8,272 B per thread |
| 155,648 | 38,912 | 34.547 ms | 644.409 ms | 18.65x | 0 / 5,744 B per thread |

## Observational end-to-end validation

These are single-concurrency, prefix-cold measurements. They are published as
observations rather than a strict A/B because the historical baseline and M16
validation profiles differ in batch size, resident-layer count, compilation
cache, and prompt seed.

| Raw context | Selector state | Client-observed TTFT | Prompt tokens / TTFT |
|---:|---|---:|---:|
| 65,536 | boundary control; already M16 before this PR | 29.893 s | 2,192.4 tok/s |
| 98,304 | historical M64 | 237.989 s | 413.1 tok/s |
| 98,304 | M16, JIT-hot unique prefix | 54.383 s | 1,807.6 tok/s |
| 155,648 | historical M64 | 730.664 s | 213.0 tok/s |
| 155,648 | M16, JIT-hot unique prefix | 110.015 s | 1,414.8 tok/s |

No patched 131,072-token full-service result is claimed or interpolated. The
README records the complete measurement boundaries, first-shape JIT samples,
and commands for independent reproduction.

## Validation

- Focused selector tests: 5 passed.
- The CUDA numerical-equivalence case is architecture-gated; it was skipped in
  the final CPU-only documentation environment.
- Ruff lint and formatting checks passed for the kernel, tests, and benchmark.
- `git diff --check` passed.
- README shell snippets and privacy checks passed.

AI assistance was used to inspect the kernel, prepare the patch and benchmark,
and draft the tests and documentation. The submitter reviewed the resulting
changes and measured results.

### CODE REVIEW

**Comment by zify9000** [comment]
Same cliff confirmed on SM86 (RTX 3080). Extending the M16 gate to the whole SM 8.x family fixes it.

Kernel benchmark (your benchmark_sm12x_mqa.py, gate relaxed to SM 8.x,
median of 5 CUDA-event samples, M16/M64 logits bitwise equal):

| Raw context | Compressed KV | M16 median | M64 median | Speedup |
| --- | --- | --- | --- | --- |
| 98,304  | 24,576 | 65.17 ms | 1457.13 ms | 22.4x |
| 131,072 | 32,768 | 61.79 ms |  363.50 ms |  5.9x |
| 155,648 | 38,912 | 62.46 ms | 1442.07 ms | 23.1x |

End-to-end (Lvllm, 2x RTX 3080, TP=2, 108K-token prompt, prefix-cold):
TTFT 650s (stock M64 selector) -> 224s JIT-cold / 207s JIT-hot (M16).
Without the patch, prefill is flat up to 64K raw context (16K compressed KV)
then collapses to ~80s per 8192-token chunk with 60s shm_broadcast stalls.

Suggest: extend the (8, 0) gate to also cover (8, 6), e.g.

    cap = current_platform.get_device_capability()
    if cap is not None and (cap.major, cap.minor) in ((8, 0), (8, 6)):
        return 16

(SM89 behavior left unchanged: the project's own 4x4090 results show no
long-context cliff, so M64 appears fine on Ada; only measured spill targets
are gated.)
